tiledb/sm/rest/rest_profile.cc

```diff
@@ -0,0 +1,230 @@
+    , old_filepath_(homedir_ + "/.tiledb/cloud.json")
+    , param_values_(default_values) {
+  // Ensure the user's $HOME is found
+  if (homedir_.empty()) {
+    throw RestProfileException(
+        "Failed to create RestProfile; $HOME is not set.");
+  }
+
+  // Fstream cannot create directories. If `homedir_/.tiledb/` DNE, create it.
+  if (!std::filesystem::exists(homedir_ + "/.tiledb")) {
+    std::filesystem::create_directory(homedir_ + "/.tiledb");
+  }
+
+  // If the local file exists, load the profile with the given name.
+  if (std::filesystem::exists(filepath_)) {
+    load_from_json_file(filepath_);
+  } else {
+    // If the old version of the file exists, load the profile from there
+    if (std::filesystem::exists(old_filepath_)) {
+      load_from_json_file(old_filepath_);
+    }
+  }
+}
+
+/* ****************************** */
+/*              API               */
+/* ****************************** */
+
+void RestProfile::set(const std::string& param, const std::string& value) {
+  // Validate incoming parameter name
+  if (default_values.count(param) == 0) {
+    throw RestProfileException(
+        "Failed to set parameter of invalid name \'" + param + "\'");
+  }
+  param_values_[param] = value;
+}
+
+std::string RestProfile::get(const std::string& param) const {
+  auto it = param_values_.find(param);
+  if (it == param_values_.end()) {
+    throw RestProfileException(
+        "Failed to retrieve parameter \'" + param + "\'");
+  }
+  return it->second;
+}
+
+/**
+ * @note The `version_` will always be listed toward the end of the local file.
+ * `nlohmann::json` does not preserve the structure of the original, top-level
+ * json object, but rather sorts its elements alphabetically.
+ * See issue [#727](https://github.com/nlohmann/json/issues/727) for details.
+ */
+void RestProfile::save() {
+  // Validate that the profile is complete (if username is set, so is password)
+  if ((param_values_["rest.username"] != RestProfile::DEFAULT_USERNAME &&
+       param_values_["rest.password"] == RestProfile::DEFAULT_PASSWORD) ||
+      (param_values_["rest.username"] == RestProfile::DEFAULT_USERNAME &&
+       param_values_["rest.password"] != RestProfile::DEFAULT_PASSWORD)) {
+    throw RestProfileException(
+        "Failed to save RestProfile; invalid username/password pairing.");
+  }
+
+  // If the file already exists, load it into a json object.
+  json data;
+  std::fstream file;
+  std::string original_filepath = filepath_;
+  if (std::filesystem::exists(filepath_)) {
+    // Temporarily append filename with a random label to guarantee atomicity.
+    filepath_ = filepath_ + random_label();
+    if (std::rename(original_filepath.c_str(), filepath_.c_str()) != 0) {
+      throw RestProfileException(
+          "Failed to save RestProfile due to an internal error.");
+    }
+
+    // Read the file into the json object.
+    file.open(filepath_, std::ofstream::in);
+    file >> data;
+    file.close();
```

Comment:
We should be declaring the `file` variables in a scope and rely in RAII to close them. Same for some subsequent occurrences.